### PR TITLE
Issues found in test cases fixed

### PR DESCRIPTION
**Pull request by UmashriR**

### CODE REVIEW

File: ZCRMiOS/Classes/RelatedListAPIHandler.swift [hunk 1]
```diff
@@ -436,6 +443,10 @@ internal class RelatedListAPIHandler : CommonAPIHandler
 	private func getZCRMAttachment(attachmentDetails : [String:Any?]) -> ZCRMAttachment
 	{
         let attachment : ZCRMAttachment = ZCRMAttachment( parentRecord : self.parentRecord )
+        if(attachmentDetails.hasValue(forKey: ResponseJSONKeys.id))
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/RelatedListAPIHandler.swift]: why did you add check here?

File: ZCRMiOS/Classes/RelatedListAPIHandler.swift [hunk 2]
```diff
@@ -481,11 +492,21 @@ internal class RelatedListAPIHandler : CommonAPIHandler
...
 	}
 	
-    private func getZCRMNote(noteDetails : [String:Any?]) -> ZCRMNote
-	{
-        let note : ZCRMNote = ZCRMNote( content : noteDetails.getString( key : ResponseJSONKeys.noteContent ) )
+    private func getZCRMNote(noteDetails : [String:Any?], note : ZCRMNote) -> ZCRMNote
+    {
         note.parentRecord = self.parentRecord
-        note.title = noteDetails.optString( key : ResponseJSONKeys.noteTitle )
+        if ( noteDetails.hasValue( forKey : ResponseJSONKeys.id ) )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/RelatedListAPIHandler.swift]: Id check is necessary?

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 3]
```diff
@@ -512,12 +512,12 @@ internal class EntityAPIHandler : CommonAPIHandler
...
     private func getTaxAsJSONArray() -> [ [ String : Any ] ]?
     {
-        if ( self.record.tax.isEmpty )
+        if ( self.record.tax == nil)
         {
             return nil
         }
         var taxJSONArray : [ [ String : Any ] ] = [ [ String : Any ] ]()
-        let allTax : [ ZCRMTax ] = self.record.tax
+        let allTax : [ ZCRMTax ] = self.record.tax!
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift]: Don't use force unwrap.

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 4]
```diff
@@ -536,12 +536,12 @@ internal class EntityAPIHandler : CommonAPIHandler
...
     private func getLineItemsAsJSONArray() -> [[String:Any]]?
     {
-        if(self.record.lineItems.isEmpty)
+        if(self.record.lineItems == nil)
         {
             return nil
         }
         var allLineItems : [[String:Any]] = [[String:Any]]()
-        let allLines : [ZCRMInventoryLineItem] = self.record.lineItems
+        let allLines : [ZCRMInventoryLineItem] = self.record.lineItems!
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift]: Don't use force unwrap.

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 5]
```diff
@@ -551,12 +551,12 @@ internal class EntityAPIHandler : CommonAPIHandler
...
     private func getPriceDetailsAsJSONArray() -> [ [ String : Any ] ]?
     {
-        if( self.record.priceDetails.isEmpty )
+        if( self.record.priceDetails == nil )
         {
             return nil
         }
         var priceDetails : [ [ String : Any ] ] = [ [ String : Any ] ]()
-        let allPriceDetails : [ ZCRMPriceBookPricing ] = self.record.priceDetails
+        let allPriceDetails : [ ZCRMPriceBookPricing ] = self.record.priceDetails!
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift]: Don't use force unwrap.

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 6]
```diff
@@ -566,12 +566,12 @@ internal class EntityAPIHandler : CommonAPIHandler
...
     private func getParticipantsAsJSONArray() -> [ [ String : Any ] ]?
     {
-        if( self.record.participants.isEmpty)
+        if( self.record.participants == nil)
         {
             return nil
         }
         var participantsDetails : [ [ String : Any ] ] = [ [ String : Any ] ]()
-        let allParticipants : [ ZCRMEventParticipant ] = self.record.participants
+        let allParticipants : [ ZCRMEventParticipant ] = self.record.participants!
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift]: Don't use force unwrap.

File: ZCRMiOS/Classes/FormDBHelper.swift [hunk 7]
```diff
@@ -228,7 +228,7 @@ public class FormDBHelper{
         contentValues.updateValue(String(fields.isPresentInQuickCreateLayout()!), forKey: ZCRMTableDetails.Fields.QUICK_CREATE_LAYOUT)
         contentValues.updateValue(fields.maxLength, forKey: ZCRMTableDetails.Fields.MAX_LENGTH)
         contentValues.updateValue(String(fields.customField), forKey: ZCRMTableDetails.Fields.CUSTOM_FIELD)
-        contentValues.updateValue(String(fields.mandatory), forKey: ZCRMTableDetails.Fields.MANDATORY)
+        contentValues.updateValue(String(fields.mandatory!), forKey: ZCRMTableDetails.Fields.MANDATORY)
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/FormDBHelper.swift]: Avoid force unwrap.

File: ZCRMiOS/Classes/ZCRMRecord.swift [hunk 8]
```diff
@@ -103,7 +104,7 @@ open class ZCRMRecord : ZCRMRecordDelegate
     /// - Parameter newLineItem: line item to be added
     public func addLineItem(newLineItem : ZCRMInventoryLineItem)
     {
-        self.lineItems.append( newLineItem )
+        self.lineItems!.append( newLineItem )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMRecord.swift]: It will crash. Avoid force unwrap.

File: ZCRMiOS/Classes/ZCRMRecord.swift [hunk 9]
```diff
@@ -120,15 +121,15 @@ open class ZCRMRecord : ZCRMRecordDelegate
     /// - Parameter priceDetail: price detail to be added
     public func addPriceDetail( priceDetail : ZCRMPriceBookPricing )
     {
-        self.priceDetails.append( priceDetail )
+        self.priceDetails!.append( priceDetail )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMRecord.swift]: It will crash. Avoid force unwrap.

File: ZCRMiOS/Classes/ZCRMRecord.swift [hunk 10]
```diff
@@ -120,15 +121,15 @@ open class ZCRMRecord : ZCRMRecordDelegate
...
     }
     
     /// Add ZCRMEventParticipant to the ZCRMRecord
     ///
     /// - Parameter participant: participant to be added
     public func addParticipant( participant : ZCRMEventParticipant )
     {
-        self.participants.append( participant )
+        self.participants!.append( participant )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMRecord.swift]: It will crash. Avoid force unwrap.

File: ZCRMiOS/Classes/ZCRMRecord.swift [hunk 11]
```diff
@@ -137,7 +138,7 @@ open class ZCRMRecord : ZCRMRecordDelegate
     public func addTax( tax : ZCRMTax )
     {
 //        self.tax[ tax.getTaxName() ] = tax
-        self.tax.append(tax)
+        self.tax!.append(tax)
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMRecord.swift]: It will crash. Avoid force unwrap.

File: ZCRMiOS/Classes/ZCRMRecord.swift [hunk 12]
```diff
@@ -103,7 +104,7 @@ open class ZCRMRecord : ZCRMRecordDelegate
     /// - Parameter newLineItem: line item to be added
     public func addLineItem(newLineItem : ZCRMInventoryLineItem)
     {
-        self.lineItems.append( newLineItem )
+        self.lineItems?.append( newLineItem )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMRecord.swift:111]: Have to initialize it before appending new element.

File: ZCRMiOS/Classes/ZCRMRecord.swift [hunk 13]
```diff
@@ -120,15 +121,15 @@ open class ZCRMRecord : ZCRMRecordDelegate
     /// - Parameter priceDetail: price detail to be added
     public func addPriceDetail( priceDetail : ZCRMPriceBookPricing )
     {
-        self.priceDetails.append( priceDetail )
+        self.priceDetails?.append( priceDetail )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMRecord.swift:132]: Have to initialize it before appending new element.

File: ZCRMiOS/Classes/ZCRMRecord.swift [hunk 14]
```diff
@@ -120,15 +121,15 @@ open class ZCRMRecord : ZCRMRecordDelegate
...
     }
     
     /// Add ZCRMEventParticipant to the ZCRMRecord
     ///
     /// - Parameter participant: participant to be added
     public func addParticipant( participant : ZCRMEventParticipant )
     {
-        self.participants.append( participant )
+        self.participants?.append( participant )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMRecord.swift:144]: Have to initialize it before appending new element.

File: ZCRMiOS/Classes/ZCRMRecord.swift [hunk 15]
```diff
@@ -137,7 +138,7 @@ open class ZCRMRecord : ZCRMRecordDelegate
     public func addTax( tax : ZCRMTax )
     {
 //        self.tax[ tax.getTaxName() ] = tax
-        self.tax.append(tax)
+        self.tax?.append(tax)
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMRecord.swift:156]: Have to initialize it before appending new element.

File: ZCRMiOS/Classes/UserAPIHandler.swift [hunk 16]
```diff
@@ -513,28 +513,25 @@ internal class UserAPIHandler : CommonAPIHandler
             throw ZCRMError.InValidError( code : ErrorCode.VALUE_NIL, message : "\( ResponseJSONKeys.countryLocale ) is must not be nil" )
         }
         user.countryLocale = userDict.getString(key : ResponseJSONKeys.countryLocale )
-        if userDict.hasValue( forKey : ResponseJSONKeys.dateFormat ) == false
+        if userDict.hasValue( forKey : ResponseJSONKeys.dateFormat )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/UserAPIHandler.swift]: This is the server field.

File: ZCRMiOS/Classes/UserAPIHandler.swift [hunk 17]
```diff
@@ -513,28 +513,25 @@ internal class UserAPIHandler : CommonAPIHandler
...
         }
-        user.country = userDict.getString( key : ResponseJSONKeys.country )
         user.fax = userDict.optString( key : ResponseJSONKeys.fax )
         if userDict.hasValue( forKey : ResponseJSONKeys.locale ) == false
         {
             throw ZCRMError.InValidError( code : ErrorCode.VALUE_NIL, message : "\( ResponseJSONKeys.locale ) is must not be nil" )
         }
         user.locale = userDict.getString( key : ResponseJSONKeys.locale )
-        if userDict.hasValue( forKey : ResponseJSONKeys.nameFormat ) == false
+        if userDict.hasValue( forKey : ResponseJSONKeys.nameFormat )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/UserAPIHandler.swift:532]: This is the server field.

File: ZCRMiOS/Classes/UserAPIHandler.swift [hunk 18]
```diff
@@ -614,15 +611,7 @@ internal class UserAPIHandler : CommonAPIHandler
...
-        let firstName = user.firstName
-        if firstName != APIConstants.STRING_MOCK
-        {
-            userJSON[ ResponseJSONKeys.firstName ] = firstName
-        }
-        else
-        {
-            userJSON[ ResponseJSONKeys.firstName ] = nil
-        }
+        userJSON[ ResponseJSONKeys.firstName ] = user.firstName
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/UserAPIHandler.swift:615]: This is the server field.

File: ZCRMiOS/Classes/UserAPIHandler.swift [hunk 19]
```diff
@@ -758,15 +667,7 @@ internal class UserAPIHandler : CommonAPIHandler
...
-        let nameFormat = user.nameFormat
-        if nameFormat != APIConstants.STRING_MOCK
-        {
-            userJSON[ ResponseJSONKeys.nameFormat ] = nameFormat
-        }
-        else
-        {
-            userJSON[ ResponseJSONKeys.nameFormat ] = nil
-        }
+        userJSON[ ResponseJSONKeys.nameFormat ] = user.nameFormat
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/UserAPIHandler.swift:671]: This is the server field.

File: ZCRMiOS/Classes/ZCRMSubformRecord.swift [hunk 20]
```diff
@@ -13,7 +13,7 @@ public class ZCRMSubformRecord : ZCRMEntity
 	public var modifiedTime : String = APIConstants.STRING_MOCK
 	public var createdTime : String = APIConstants.STRING_MOCK
     var layout : ZCRMLayout?
-	public var fieldNameVsValue : [String:Any?] = [String:Any?]()
+	public var fieldNameVsValue : [String:Any] = [String:Any]()
```
Comment (vijayakrishnap) [inline ZCRMiOS/Classes/ZCRMSubformRecord.swift:16]: Any? vs Any while getting values will this work fine?

File: ZCRMiOS/Classes/ZCRMTag.swift [hunk 21]
```diff
@@ -12,13 +12,27 @@ open class ZCRMTag : ZCRMTagDelegate
...
         super.init( tagId : APIConstants.INT64_MOCK, tagName : tagName, moduleAPIName : APIConstants.STRING_MOCK )
     }
     
-    init()
+    internal init()
     {
         super.init( tagId : APIConstants.INT64_MOCK, moduleAPIName : APIConstants.STRING_MOCK)
     }
+    
+    public func update( updateTag : ZCRMTag, completion : @escaping ( Result.DataResponse< ZCRMTag, APIResponse > ) -> () )
```
Comment (vijayakrishnap) [inline ZCRMiOS/Classes/ZCRMTag.swift]: Whats this update method? if its update of tag, why does it require an argument?

File: ZCRMiOS/Classes/ZCRMTimelineEvent.swift [hunk 22]
```diff
@@ -0,0 +1,33 @@
+//
+//  ZCRMTimeline.swift
+//  ZCRMiOS
+//
+//  Created by Umashri R on 26/09/18.
+//
+
+open class ZCRMTimeline : ZCRMEntity
```
Comment (vijayakrishnap) [inline ZCRMiOS/Classes/ZCRMTimelineEvent.swift]: Remove this from this merge.

**Review by vijayakrishnap — CHANGES_REQUESTED** [review]

File: ZCRMiOS/Classes/ZCRMSubformRecord.swift (hunk 20 above)
Comment (UmashriR) [reply ZCRMiOS/Classes/ZCRMSubformRecord.swift:16]: Yes. Any can have optional values.

**Review by vijayakrishnap — APPROVED** [review]
